### PR TITLE
Travis build changes

### PR DESCRIPTION
**Pull request by posix4e**
```
Prepare seperate environments for different operating systems.
Generate libpq.rs rather than checking it in
Add basic postgres integration tests
```

### CODE REVIEW

File: generate_bindings.sh [hunk 1]
```diff
@@ -9,10 +11,22 @@ echo '#include "fmgr.h"' >> /tmp/postgres.c
 echo '#include "replication/output_plugin.h"' >> /tmp/postgres.c
 echo '#include "replication/logical.h"' >> /tmp/postgres.c
 
+gcc -I /usr/local/Cellar/postgresql/9.5.0/ \
```
Comment (solidsnack) [inline generate_bindings.sh]: This stuff can all be gathered from `pg_config`; that would seem to be a better investment of effort.

File: src/lib.rs [hunk 2]
```diff
@@ -35,11 +25,11 @@ pub type LogicalOutputPluginInit =
                               -> ()>;
 */
 
-extern fn startup(ctx: *mut libpq::Struct_LogicalDecodingContext,
+unsafe extern fn startup(ctx: *mut libpq::Struct_LogicalDecodingContext,
```
Comment (solidsnack) [inline src/lib.rs:28]: I don't think this works here: `unsafe extern` and `extern` are not the same; so if the earlier types work, I'm not sure how these can. Does it have to do with changes in bindgen?

File: ci_tools/prepare_linux.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
```
Comment (solidsnack) [inline ci_tools/prepare_linux.sh:1]: We should probably leave the system installation of Postgres alone, on both OS X and Linux. Using a temporary database directory is much more congenial to the development habits of many, who would expect to be able to run tests safely from their laptop.

File: ci_tools/prepare_linux.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+set -ex
+VERSION=9.4
+sudo pg_ctlcluster $VERSION main stop
+make clean all 
+sudo /bin/mkdir -p /usr/share/postgresql/${VERSION}/extension
+sudo /bin/mkdir -p /usr/share/postgresql/${VERSION}/extension
+sudo /bin/mkdir -p /usr/lib/postgresql/${VERSION}/lib
+sudo /bin/mkdir -p /usr/share/doc/postgresql-doc-${VERSION}/extension
```
Comment (solidsnack) [inline ci_tools/prepare_linux.sh:9]: This stuff should all be handled transparently by `make install`.

**Comment by solidsnack** [comment]
There is a lot of stuff going on in this PR.

The testing, the binding generation and the new code could all be separated from each other.

You are really pushing hard for generating `libpq.rs` but it's not clear there is any value in that in the near term.

I would like to be able to merge testing stuff apart from all that. Please consider breaking this up.

File: generate_bindings.sh (hunk 1 above)
Comment (posix4e) [inline generate_bindings.sh]: Fixed in the new commit

File: src/lib.rs (hunk 2 above)
Comment (posix4e) [inline src/lib.rs:28]: Yea it does alas

File: ci_tools/prepare_linux.sh (hunk 3 above)
Comment (posix4e) [inline ci_tools/prepare_linux.sh:1]: Can we do that in a later commit. These are just for travis and the ci environment.

File: ci_tools/prepare_linux.sh (hunk 4 above)
Comment (posix4e) [inline ci_tools/prepare_linux.sh:9]: I would love to, alas, make install is currently broken.

**Comment by posix4e** [comment]
Generating the h file is important if we wanna support linking across multiple versions.
At least that's what https://travis-ci.org/posix4e/jsoncdc/jobs/106252238 suggests when I tried to do this. osx7 has them broken out

File: ci_tools/prepare_linux.sh (hunk 3 above)
Comment (posix4e) [inline ci_tools/prepare_linux.sh:1]: Should I make it clear a developer should not run it. That's why it's in the ci_tools directory

File: ci_tools/prepare_linux.sh (hunk 4 above)
Comment (posix4e) [inline ci_tools/prepare_linux.sh:9]: https://github.com/posix4e/jsoncdc/issues/30

**Comment by posix4e** [comment]
Closing this. We discuse

**Pull request closed without merging**